scalar/uint.go

```diff
@@ -58,121 +71,167 @@ func (s *Uint64) Set(val any) error {
 	switch value := val.(type) {
 	case int8:
 		if value < 0 {
-			return &ValidationError{Type: arrow.PrimitiveTypes.Uint64, Msg: "int8 less than 0", Value: value}
+			return &ValidationError{Type: s.DataType(), Msg: "int8 less than 0", Value: value}
 		}
-		s.Value = uint64(value)
+		return s.Set(uint64(value))
 	case int16:
 		if value < 0 {
-			return &ValidationError{Type: arrow.PrimitiveTypes.Uint64, Msg: "int16 less than 0", Value: value}
+			return &ValidationError{Type: s.DataType(), Msg: "int16 less than 0", Value: value}
 		}
-		s.Value = uint64(value)
+		return s.Set(uint64(value))
 	case int32:
 		if value < 0 {
-			return &ValidationError{Type: arrow.PrimitiveTypes.Uint64, Msg: "int32 less than 0", Value: value}
+			return &ValidationError{Type: s.DataType(), Msg: "int32 less than 0", Value: value}
 		}
-		s.Value = uint64(value)
+		return s.Set(uint64(value))
 	case int64:
 		if value < 0 {
-			return &ValidationError{Type: arrow.PrimitiveTypes.Uint64, Msg: "int64 less than 0", Value: value}
+			return &ValidationError{Type: s.DataType(), Msg: "int64 less than 0", Value: value}
 		}
-		s.Value = uint64(value)
+		return s.Set(uint64(value))
 	case int:
 		if value < 0 {
-			return &ValidationError{Type: arrow.PrimitiveTypes.Uint64, Msg: "int less than 0", Value: value}
+			return &ValidationError{Type: s.DataType(), Msg: "int less than 0", Value: value}
 		}
-		s.Value = uint64(value)
+		return s.Set(uint64(value))
 	case uint8:
-		s.Value = uint64(value)
+		return s.Set(uint64(value))
 	case uint16:
-		s.Value = uint64(value)
+		return s.Set(uint64(value))
 	case uint32:
-		s.Value = uint64(value)
+		return s.Set(uint64(value))
 	case uint64:
+		if err := s.validateValue(value); err != nil {
+			return err
+		}
 		s.Value = value
 	case uint:
-		s.Value = uint64(value)
+		return s.Set(uint64(value))
 	case float32:
 		if value < 0 {
-			return &ValidationError{Type: arrow.PrimitiveTypes.Uint64, Msg: "float32 less than 0", Value: value}
+			return &ValidationError{Type: s.DataType(), Msg: "float32 less than 0", Value: value}
 		}
-		s.Value = uint64(value)
+		return s.Set(uint64(value))
```

Comment:
check for `math.MaxUint64`?

Reply:
maxuint64 is bigger than maxfloat32 (or maxfloat64) so no need to check here. uint64 branch calls `validateValue` to check against target bitwidth.

Reply:
https://go.dev/play/p/5au3XVMfX4C